src/util/buf.rs

```diff
@@ -11,93 +11,429 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::io::{Result, Write};
-use std::collections::VecDeque;
+use std::io::{Result, Write, Read};
+use std::fmt::{self, Debug, Formatter};
+use alloc::raw_vec::RawVec;
+use std::{cmp, ptr, slice, mem};
 
 use bytes::{ByteBuf, MutByteBuf, alloc};
 pub use mio::{TryRead, TryWrite};
 
+use util::escape;
+
 // `create_mem_buf` creates the buffer with fixed capacity s.
 pub fn create_mem_buf(s: usize) -> MutByteBuf {
     unsafe {
         ByteBuf::from_mem_ref(alloc::heap(s.next_power_of_two()), s as u32, 0, s as u32).flip()
     }
 }
 
-pub struct SendBuffer {
-    buf: VecDeque<u8>,
+/// `PipeBuffer` is useful when you want to move data from `Write` to a `Read` or vice versa.
+pub struct PipeBuffer {
+    // the index of the first byte of written data.
+    start: usize,
+    // the index of buf that new data should be written in.
+    end: usize,
+    buf: RawVec<u8>,
 }
 
-impl SendBuffer {
-    pub fn new(n: usize) -> SendBuffer {
-        SendBuffer { buf: VecDeque::with_capacity(n) }
+impl PipeBuffer {
+    pub fn new(capacity: usize) -> PipeBuffer {
+        PipeBuffer {
+            start: 0,
+            end: 0,
+            // one extra byte to indicate if buf is full or empty.
+            buf: RawVec::with_capacity(capacity + 1),
```

Comment:
It's kind of magic, Could you use another field to indicate whether buf is full or empty ?